File: Pandacap/Podcasts/PodcastStream.cs
using System.Net.Http.Headers;

namespace Pandacap.Podcasts
{
    internal class PodcastStream(
        IHttpClientFactory httpClientFactory,
        Uri Uri,
        long ContentLength) : Stream
    {
        public override bool CanRead => true;

        public override bool CanSeek => true;

        public override bool CanWrite => false;

        public override long Length => ContentLength;

        public override long Position { get; set; }

        public override void Flush() { }

        public override int Read(byte[] buffer, int offset, int count) =>
            throw new NotImplementedException();

        public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
        {
            long to = Math.Min(
                Position + buffer.Length,
                Length);

            if (Position == to)
                return 0;

            using var client = httpClientFactory.CreateClient();

            using var req = new HttpRequestMessage(HttpMethod.Get, Uri);
            req.Headers.Range = new RangeHeaderValue(Position, to);

            using var resp = await client.SendAsync(req, cancellationToken);
            resp.EnsureSuccessStatusCode();

            using var stream = await resp.Content.ReadAsStreamAsync(cancellationToken);

            int read = await stream.ReadAsync(buffer, cancellationToken);
            Position += read;

            return read;
        }

        public override long Seek(long offset, SeekOrigin origin)
        {
            return Position = origin switch
            {
                SeekOrigin.Begin => offset,
                SeekOrigin.Current => Position + offset,
                SeekOrigin.End => ContentLength + offset,
                _ => throw new ArgumentException($"Unrecgonized origin {origin}", nameof(origin))
            };
        }

        public override void SetLength(long value) =>
            throw new NotImplementedException();

        public override void Write(byte[] buffer, int offset, int count) =>
            throw new NotImplementedException();
    }
}

File: Pandacap.Podcasts/PodcastStream.fs
namespace Pandacap.Podcasts

open System
open System.IO
open System.Net.Http
open System.Net.Http.Headers
open System.Threading.Tasks

type internal PodcastStream(
    httpClientFactory: IHttpClientFactory,
    uri: Uri,
    contentLength: int64
) =
    inherit Stream()

    let mutable pos = 0L

    override _.CanRead = true
    override _.CanSeek = true
    override _.CanWrite = false

    override _.Length = contentLength

    override _.Position with get () = pos and set v = pos <- v

    override _.Flush() = ()

    override _.Read(_, _, _) =  raise (NotImplementedException())

    override _.ReadAsync(buffer, cancellationToken) = ValueTask<int> (task {
        let posTo = min (pos + int64 buffer.Length) contentLength

        if posTo = pos then
            return 0
        else
            use client = httpClientFactory.CreateClient()

            use req = new HttpRequestMessage(HttpMethod.Get, uri)
            req.Headers.Range <- new RangeHeaderValue(pos, posTo)

            use! resp = client.SendAsync(req, cancellationToken)
            use! stream = resp.EnsureSuccessStatusCode().Content.ReadAsStreamAsync(cancellationToken)

            let! read = stream.ReadAsync(buffer, cancellationToken)
            pos <- pos + int64 read

            return read
    })

    override _.Seek(offset, origin) =
        pos <-
            match origin with
            | SeekOrigin.Begin -> offset
            | SeekOrigin.Current -> pos + offset
            | SeekOrigin.End -> contentLength + offset
            | x -> raise (ArgumentException($"Unrecgonized origin {x}", nameof origin))
        pos

    override _.SetLength(_) = raise (NotSupportedException())

    override _.Write(_, _, _) = raise (NotSupportedException())
